initgeneration/initgeneration.go

```diff
@@ -0,0 +1,247 @@
+func (g *InitGenerator) GenerateForNamespace(ctx context.Context, dkName, targetNs string) error {
+	g.logger.Info("Reconciling namespace init secret for", "namespace", targetNs)
+	var dk dynatracev1alpha1.DynaKube
+	if err := g.client.Get(context.TODO(), client.ObjectKey{Name: dkName, Namespace: g.namespace}, &dk); err != nil {
+		return err
+	}
+	data, err := g.generate(ctx, &dk)
+	if err != nil {
+		return err
+	}
+	return kubeobjects.CreateOrUpdateSecretIfNotExists(g.client, g.apiReader, webhook.SecretConfigName, targetNs, data, corev1.SecretTypeOpaque, g.logger)
+}
+
+// GenerateForDynakube creates/updates the init secret for EVERY namespace for the given dynakube.
+// Used by the dynakube controller during reconcile.
+func (g *InitGenerator) GenerateForDynakube(ctx context.Context, dk *dynatracev1alpha1.DynaKube) (bool, error) {
+	g.logger.Info("Reconciling namespace init secret for", "dynakube", dk.Name)
+	data, err := g.generate(ctx, dk)
+	if err != nil {
+		return false, err
+	}
+
+	nsList, err := mapper.GetNamespacesForDynakube(ctx, g.apiReader, dk.Name)
+	if err != nil {
+		return false, err
+	}
+	for _, targetNs := range nsList {
+		g.logger.Info("Updating init secret from dynakube for", "namespace", targetNs.Name)
+		if err = kubeobjects.CreateOrUpdateSecretIfNotExists(g.client, g.apiReader, webhook.SecretConfigName, targetNs.Name, data, corev1.SecretTypeOpaque, g.logger); err != nil {
+			return false, err
+		}
+	}
+	g.logger.Info("Done updating init secrets")
+	return true, nil
+}
+
+// generate gets the necessary info the create the init secret data
+func (g *InitGenerator) generate(ctx context.Context, dk *dynatracev1alpha1.DynaKube) (map[string][]byte, error) {
+	kubeSystemUID, err := kubesystem.GetUID(g.apiReader)
+	if err != nil {
+		return nil, err
+	}
+
+	infraMonitoringNodes, err := g.getInfraMonitoringNodes(dk)
+	if err != nil {
+		return nil, err
+	}
+
+	script, err := g.prepareScriptForDynaKube(dk, kubeSystemUID, infraMonitoringNodes)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := script.generate()
+	if err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func (g *InitGenerator) prepareScriptForDynaKube(dk *dynatracev1alpha1.DynaKube, kubeSystemUID types.UID, infraMonitoringNodes map[string]string) (*script, error) {
+	var tokens corev1.Secret
+	if err := g.client.Get(context.TODO(), client.ObjectKey{Name: dk.Tokens(), Namespace: g.namespace}, &tokens); err != nil {
+		return nil, errors.WithMessage(err, "failed to query tokens")
+	}
+
+	var proxy string
+	if dk.Spec.Proxy != nil {
+		if dk.Spec.Proxy.ValueFrom != "" {
+			var ps corev1.Secret
+			if err := g.client.Get(context.TODO(), client.ObjectKey{Name: dk.Spec.Proxy.ValueFrom, Namespace: g.namespace}, &ps); err != nil {
+				return nil, fmt.Errorf("failed to query proxy: %w", err)
+			}
+			proxy = string(ps.Data[proxyInitSecretField])
+		} else if dk.Spec.Proxy.Value != "" {
+			proxy = dk.Spec.Proxy.Value
+		}
+	}
```

Comment:
This code appears frequently in the code base, consider refactoring it to e.g. kubeobjects